core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,385 @@
+      }
+
+      val maxTaskFailures = conf.getInt("spark.task.maxFailures", 4)
+      val maxNodeAttempts = conf.get(config.MAX_TASK_ATTEMPTS_PER_NODE)
+
+      if (maxTaskFailures <= maxNodeAttempts) {
+        throw new IllegalArgumentException(s"${config.MAX_TASK_ATTEMPTS_PER_NODE.key} " +
+          s"( = ${maxNodeAttempts}) was <= spark.task.maxFailures " +
+          s"( = ${maxTaskFailures} ).  Though blacklisting is enabled, with this configuration, " +
+          s"Spark will not be robust to one failed disk.  Increase " +
+          s"${config.MAX_TASK_ATTEMPTS_PER_NODE.key} or spark.task.maxFailures, or disable " +
+          s"blacklisting with ${config.BLACKLIST_ENABLED.key}")
+      }
+    }
+
+  }
+}
+
+/** Failures for one executor, within one taskset */
+private[scheduler] final class ExecutorFailuresInTaskSet(val node: String) {
+  /**
+   * Mapping from index of the tasks in the taskset, to the number of times it has failed on this
+   * executor.
+   */
+  val taskToFailureCountAndExpiryTime = HashMap[Int, (Int, Long)]()
+  def updateWithFailure(taskIndex: Int, failureExpiryTime: Long): Unit = {
+    val (prevFailureCount, prevFailureExpiryTime) =
+      taskToFailureCountAndExpiryTime.getOrElse(taskIndex, (0, -1L))
+    assert(failureExpiryTime >= prevFailureExpiryTime)
+    taskToFailureCountAndExpiryTime(taskIndex) = (prevFailureCount + 1, failureExpiryTime)
+  }
+  def numUniqueTasksWithFailures: Int = taskToFailureCountAndExpiryTime.size
+
+
+  override def toString(): String = {
+    s"numUniqueTasksWithFailures= $numUniqueTasksWithFailures; " +
+      s"tasksToFailureCount = $taskToFailureCountAndExpiryTime"
+  }
+}
+
+/**
+ * Tracks all failures for one executor (that have not passed the timeout).  Designed to efficiently
+ * remove failures that are older than the timeout, and query for the number of unique failed tasks.
+ */
+private[scheduler] final class ExecutorFailureList extends Logging {
+
+  private case class TaskId(stage: Int, stageAttempt: Int, taskIndex: Int)
+
+  /**
+   * All failures on this executor in successful task sets, sorted by time ascending.
+   */
+  private var failures = ArrayBuffer[(TaskId, Long)]()
+
+  def addFailures(
+      stage: Int,
+      stageAttempt: Int,
+      failuresInTaskSet: ExecutorFailuresInTaskSet): Unit = {
+    // The new failures may interleave with the old ones, so rebuild the failures in sorted order.
+    // This shouldn't be expensive because if there were a lot of failures, the executor would
+    // have been blacklisted.
+    if (failuresInTaskSet.taskToFailureCountAndExpiryTime.nonEmpty) {
+      failuresInTaskSet.taskToFailureCountAndExpiryTime.foreach { case (taskIdx, (_, time)) =>
+        failures += ((TaskId(stage, stageAttempt, taskIdx), time))
+      }
+      // sort by failure time, so we can quickly determine if any failure has gone past the timeout
+      failures = failures.sortBy(_._2)
+    }
+  }
+
+  def minExpiryTime: Long = failures.head._2
```

Comment:
I'm seeing an exception on this when a node is unblacklisted after the timeout. In this case it blacklisted a set of executors and nodes and then after a 10 minutes timeout went to unblacklist.

java.util.NoSuchElementException: next on empty iterator
        at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
        at scala.collection.Iterator$$anon$2.next(Iterator.scala:37)
        at scala.collection.IndexedSeqLike$Elements.next(IndexedSeqLike.scala:63)
        at scala.collection.IterableLike$class.head(IterableLike.scala:107)
        at scala.collection.mutable.ArrayBuffer.scala$collection$IndexedSeqOptimized$$super$head(ArrayBuffer.scala:48)
        at scala.collection.IndexedSeqOptimized$class.head(IndexedSeqOptimized.scala:126)
        at scala.collection.mutable.ArrayBuffer.head(ArrayBuffer.scala:48)
        at org.apache.spark.scheduler.ExecutorFailureList.minExpiryTime(BlacklistTracker.scala:305)
        at org.apache.spark.scheduler.ExecutorFailureList.dropFailuresWithTimeoutBefore(BlacklistTracker.scala:314)
        at org.apache.spark.scheduler.BlacklistTracker$$anonfun$applyBlacklistTimeout$1.apply(BlacklistTracker.scala:100)
        at org.apache.spark.scheduler.BlacklistTracker$$anonfun$applyBlacklistTimeout$1.apply(BlacklistTracker.scala:99)
        at scala.collection.mutable.HashMap$$anon$2$$anonfun$foreach$3.apply(HashMap.scala:108)
        at scala.collection.mutable.HashMap$$anon$2$$anonfun$foreach$3.apply(HashMap.scala:108)
        at scala.collection.mutable.HashTable$class.foreachEntry(HashTable.scala:230)
        at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
        at scala.collection.mutable.HashMap$$anon$2.foreach(HashMap.scala:108)
        at org.apache.spark.scheduler.BlacklistTracker.applyBlacklistTimeout(BlacklistTracker.scala:99)
        at org.apache.spark.scheduler.TaskSchedulerImpl$$anonfun$resourceOffers$2.apply(TaskSchedulerImpl.scala:331)
        at org.apache.spark.scheduler.TaskSchedulerImpl$$anonfun$resourceOffers$2.apply(TaskSchedulerImpl.scala:331)
        at scala.Option.foreach(Option.scala:257)
        at org.apache.spark.scheduler.TaskSchedulerImpl.resourceOffers(TaskSchedulerImpl.scala:331)
        at org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend$DriverEndpoint.org$apache$spark$scheduler$cluster$CoarseGrainedSchedulerBackend$DriverEndpoint$$makeOffers(CoarseGrainedSchedulerBackend.scala:217)
        at org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend$DriverEndpoint$$anonfun$receive$1.applyOrElse(CoarseGrainedSchedulerBackend.scala:135)
        at org.apache.spark.rpc.netty.Inbox$$anonfun$process$1.apply$mcV$sp(Inbox.scala:117)
        at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:205)
        at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:101)
        at org.apache.spark.rpc.netty.Dispatcher$MessageLoop.run(Dispatcher.scala:213)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)